wlroots/xwayland.py

```diff
@@ -0,0 +1,462 @@
+# Copyright (c) 2022 Matt Colligan
+
+from __future__ import annotations
+
+import enum
+from typing import TYPE_CHECKING
+
+from pywayland.server import Signal
+
+from wlroots import ffi, lib, Ptr, PtrHasData, str_or_none
+from wlroots.wlr_types.surface import Surface as WlrSurface
+
+if TYPE_CHECKING:
+    from pywayland.server import Display
+
+    from wlroots.wlr_types import Compositor, Seat
+    from wlroots.wlr_types.xdg_shell import SurfaceCallback, T
+
+
+class ServerOptions(Ptr):
+    def __init__(self, ptr) -> None:
+        self._ptr = ptr
+
+    @classmethod
+    def new(
+        cls, lazy: bool, enable_wm: bool, no_touch_pointer_emulation: bool
+    ) -> ServerOptions:
+        ptr = ffi.new("struct wlr_xwayland_server_options *")
+        ptr.lazy = lazy
+        ptr.enable_wm = enable_wm
+        ptr.no_touch_pointer_emulation = no_touch_pointer_emulation
+        return cls(ptr)
+
+    @property
+    def lazy(self) -> bool:
+        return self._ptr.lazy
+
+    @lazy.setter
+    def lazy(self, lazy: bool) -> None:
+        self._ptr.lazy = lazy
+
+    @property
+    def enable_wm(self) -> bool:
+        return self._ptr.enable_wm
+
+    @enable_wm.setter
+    def enable_wm(self, enable_wm: bool) -> None:
+        self._ptr.enable_wm = enable_wm
+
+    @property
+    def no_touch_pointer_emulation(self) -> bool:
+        return self._ptr.no_touch_pointer_emulation
+
+    @no_touch_pointer_emulation.setter
+    def no_touch_pointer_emulation(self, no_touch_pointer_emulation: bool) -> None:
+        self._ptr.no_touch_pointer_emulation = no_touch_pointer_emulation
+
+
+class Server(PtrHasData):
+    def __init__(self, ptr) -> None:
+        self._ptr = ptr
+
+        self.ready_event = Signal(ptr=ffi.addressof(self._ptr.events.ready))
+        self.destroy_event = Signal(ptr=ffi.addressof(self._ptr.events.destroy))
+
+    @classmethod
+    def create(cls, display: Display, options: ServerOptions) -> Server | None:
+        ptr = lib.wlr_xwayland_server_create(display._ptr, options._ptr)
+        if ptr == ffi.NULL:
+            return None
+        return cls(ptr)
+
+    def destroy(self) -> None:
+        lib.wlr_xwayland_server_destroy(self._ptr)
```

Comment:
Can this be called multiple times? Should we null out `self._ptr` when this is called? Does this need to be called when the ptr is created from `wlr_xwayland_server_create` (that is, should we add this as the `ffi.gc` action in `create`)?